Gems/AtomTressFX/Code/Builders/HairAssetBuilder.cpp

```diff
@@ -0,0 +1,168 @@
+                AzFramework::StringFunc::Path::GetFullFileName(request.m_fullPath.c_str(), tfxFileName);
+
+                // Create the path to the result .tfxhair file inside the tempDirPath.
+                AZStd::string destPath;
+                AzFramework::StringFunc::Path::ConstructFull(request.m_tempDirPath.c_str(), tfxFileName.c_str(), destPath, true);
+                AZ::StringFunc::Path::ReplaceExtension(destPath, AMD::TFXCombinedFileExtension);
+
+                // Create and open the .tfxhair we are writing to. 
+                AZ::IO::FileIOStream outStream(destPath.data(), AZ::IO::OpenMode::ModeWrite | AZ::IO::OpenMode::ModeCreatePath);
+                if (!outStream.IsOpen())
+                {
+                    AZ_TracePrintf(
+                        AssetBuilderSDK::ErrorWindow, "Error: Failed job %s because .tfxhair file cannot be created.\n", request.m_fullPath.c_str());
+                    response.m_resultCode = AssetBuilderSDK::ProcessJobResult_Failed;
+                    return;
+                }
+
+                // Write the .tfxhair file header as a placeholder
+                AMD::TressFXCombinedHairFileHeader header; 
+                outStream.Write(sizeof(AMD::TressFXCombinedHairFileHeader), &header);
+
+                // Combine .tfx, .tfxbone, .tfxmesh file into .tfxhair file
+                auto writeToStream = [](const AZStd::string& fullpath, AZ::IO::FileIOStream& outStream, bool required) {
+                    IO::FileIOStream inStream;
+                    if (!inStream.Open(fullpath.c_str(), IO::OpenMode::ModeRead))
+                    {
+                        if (required)
+                        {
+                            AZ_TracePrintf(AssetBuilderSDK::ErrorWindow,
+                                "Error: Failed job %s because the file is either missing or cannot be opened.\n", fullpath.c_str());
+                        }
+                        return (AZ::IO::SizeType)0;
+                    }
+                    const AZ::IO::SizeType dataSize = inStream.GetLength();
+                    AZStd::vector<AZ::u8> fileBuffer(dataSize);
+                    inStream.Read(dataSize, fileBuffer.data());
+                    return outStream.Write(dataSize, fileBuffer.data());
+                };
+
+                // Write .tfx file to the combined .tfxhair file.
+                AZStd::string sourcePath = request.m_fullPath;
+                const AZ::IO::SizeType tfxSize = writeToStream(sourcePath, outStream, true);
+
+                // Move on to .tfxbone file.
+                AZ::StringFunc::Path::ReplaceExtension(sourcePath, AMD::TFXBoneFileExtension);
+                const AZ::IO::SizeType tfxBoneSize = writeToStream(sourcePath, outStream, true);
+
+                // Move on to .tfxmesh file.
+                AZ::StringFunc::Path::ReplaceExtension(sourcePath, AMD::TFXMeshFileExtension);
+                writeToStream(sourcePath, outStream, false);
```

Comment:
again, on the topic of traceprintfs, do you have enough forensic traceprintfs such that if any of these steps failed or had a problem you'd be able to see in the log what went wrong?

Reply:
added quite a few - tnx for noticing.  @amzn-rhhong, please review after as this is mainly your file.